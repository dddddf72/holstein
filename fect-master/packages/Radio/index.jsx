import { computed, onMounted, ref, toRefs, watch, watchEffect } from 'vue'
import { createNameSpace, theme, validator,useProvider } from '../utils'

const { normalSizes } = theme

const [createComponent] = createNameSpace('Radio')

const READNONLY_RADIO_GROUP_KEY = 'radioGroupKey'

import './radio.less'


const queryRadioSize = (radioSize)=>{
  const size = {
    mini:'12px',
    small:'14px',
    medium:'16px',
    large:'18px',
  }
  return size[radioSize]
}


export default createComponent({
  props: {
    checked: Boolean,
    disabled:Boolean,
    value: { type: [String,Number] ,required:true },
    size: {
      type: String,
      validator: validator.enums(normalSizes),
      default: 'medium',
    },
  },
  emits: ['change'],
  setup(props, { attrs, slots, emit }) {
    
    const { checked, value, size ,disabled } = toRefs(props)
    const { ctx } = useProvider(READNONLY_RADIO_GROUP_KEY)
    const radioValue = ref(value?.value) //single radio value
    const radioSize = ref(size.value)
    const isDisabled = ref(disabled.value)
    const selfChecked = ref(!!checked.value)

    const changeStatus = ()=>{
      isDisabled.value = ctx.props.disabled
      radioSize.value = ctx.props.size
      /** refactor */
      const parentValue = ctx.props?.initialValue || null
      if (ctx.groupValue.value) selfChecked.value = ctx.groupValue.value === value.value
      if (!ctx.groupValue.value){
        if (parentValue) selfChecked.value = parentValue === value.value
      }
    }

    /**
     * when component init,it will execute once.
     * watchEffect will executed auto when the dependence is changed 
     */
    if (ctx){
      watchEffect(()=>{
        changeStatus()
      })
    } 
    /**
     * without radioGroup 
     */
    if (!ctx){
      watchEffect(()=>{
        selfChecked.value = checked.value ? true : false
      })
    }
   
    const handlerChange = (e) => {
      if (isDisabled.value) return
      selfChecked.value = !selfChecked.value
      const radioEvent = {
        target:{
          checked:selfChecked.value,
        },
        stopPropagation: e.stopPropagation,
        preventDefault: e.preventDefault,
        nativeEvent: e,
      }
      if (ctx){
        ctx.groupValue.value = value.value
        ctx.updateState && ctx.updateState(radioValue.value)
      }
      if (!ctx){
        emit('change', radioEvent)
      }
    }

    const calcRadioSize = computed(()=>{
      const _size = queryRadioSize(radioSize.value)
      const style = {}
      style['--radioSize'] = _size
      return style
    })

    return () => (
      <div 
        {...attrs}
        className={`fect-radio ${attrs?.class ? attrs.class : ''}`} style={calcRadioSize.value}>
        <label className={` ${isDisabled.value ? 'disabled' : ''}`}>
          <input
            type="radio"
            value={radioValue.value}
            checked={selfChecked.value}
            onChange={handlerChange}
            disabled={isDisabled.value}
          ></input>
          <span className={'fect-radio-name'}>
            <span 
              className={`fect-radio-point ${isDisabled.value ? 'disabled' : ''} ${selfChecked.value ? 'active' : ''}`}  />
            {slots.default?.()}
          </span>
        
        </label>
      </div>
    )
  },
})
