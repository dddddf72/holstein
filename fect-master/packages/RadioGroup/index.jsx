import { computed, toRefs ,ref  } from 'vue'
import { createNameSpace, theme, validator,createProvider } from '../utils'

const { normalSizes } = theme

const [createComponent] = createNameSpace('RadioGroup')

const READNONLY_RADIO_GROUP_KEY = 'radioGroupKey'

import './radioGroup.less'

export default createComponent({
  props: {
    initialValue: [String, Number],
    useRow: Boolean,
    disabled: Boolean,
    size: {
      type: String,
      validator: validator.enums(normalSizes),
      default: 'medium',
    },
  },
  emits: ['change'],
  setup(props, { attrs, slots,emit }) {
    const {  useRow } = toRefs(props)
    const groupValue = ref(null)
    const updateState = (nextVal)=> emit('change',nextVal)
    
    const { provider } = createProvider(READNONLY_RADIO_GROUP_KEY)
    provider({ props , updateState ,groupValue })
    return () => (
      <>
        <div 
          {...attrs} 
          className={`fect-radio-group 
            ${useRow.value ? 'useRow' : ''}
            ${attrs?.class ? attrs.class : ''}`}>
          {slots.default?.()}
        </div>
        
      </>
    )
  },
})
