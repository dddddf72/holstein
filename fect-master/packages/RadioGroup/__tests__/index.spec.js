import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import RadioGroup from '../index'
import Radio from '../../Radio'

describe('RadioGroup', () => {
  it('should be render as element', () => {
    const radioGroup = mount(RadioGroup)
    expect(radioGroup.html()).toMatchSnapshot()
    expect(() => radioGroup.unmount()).not.toThrow()
  })

  it('should be support initialValue', () => {
    const wrapper = mount(
      <RadioGroup initialValue={1}>
        <Radio value={1}></Radio>
        <Radio value={2}></Radio>
      </RadioGroup>,
    )

    expect(wrapper.html()).toMatchSnapshot()
    const els = wrapper.findAll('.fect-radio-point')
    expect(els[0].classes('active')).toBeTruthy()
  })

  it('should be support useRow', () => {
    const wrapper = mount(<RadioGroup useRow={true}></RadioGroup>)
    expect(wrapper.find('.fect-radio-group').classes('useRow')).toBe(true)
  })

  it('should be support disabled', () => {
    const wrapper = mount(
      <RadioGroup disabled={true}>
        <Radio value={1} />
        <Radio value={2} />
      </RadioGroup>,
    )

    const labels = wrapper.findAll('label')
    expect(labels[0].classes('disabled')).toBeTruthy()
    expect(labels[1].classes('disabled')).toBeTruthy()
  })

  it('should be support differentSizes', () => {
    const wrapper = mount({
      setup() {
        return { sizeList: ['mini', 'small', 'medium', 'large'] }
      },
      render(h) {
        return (
          <>
            {this.sizeList.map((item) => (
              <RadioGroup size={item} key={item}>
                <Radio value={item} />
              </RadioGroup>
            ))}
          </>
        )
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should be support emit change', async () => {
    const wrapper = mount({
      emits: ['change'],
      setup(props, { emit }) {
        const test = ref(0)
        const handlerChange = (next) => {
          test.value = next
          emit('change', next)
        }
        return {
          test,
          handlerChange,
        }
      },
      render() {
        return (
          <>
            <RadioGroup onChange={this.handlerChange}>
              <Radio value={1} />
              <Radio value={2} />
            </RadioGroup>
          </>
        )
      },
    })

    const labels = wrapper.findAll('[type="radio"]')
    await labels[0].trigger('change')
    expect(wrapper.vm.test).toEqual(1)
    expect(wrapper.emitted().change[0]).toEqual([1])
    await labels[1].trigger('change')
    expect(wrapper.vm.test).toEqual(2)
    // expect(wrapper.emitted().change[2]).toEqual([2])
  })
})
