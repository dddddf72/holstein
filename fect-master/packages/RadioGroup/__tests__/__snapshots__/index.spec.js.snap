// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup should be render as element 1`] = `
<div class="fect-radio-group 
            
            ">
  <!---->
</div>
`;

exports[`RadioGroup should be support differentSizes 1`] = `
<div class="fect-radio-group 
            
            ">
  <div class="fect-radio " style="--radioSize: 12px;"><label class=" "><input type="radio" value="mini"><span class="fect-radio-name"><span class="fect-radio-point  "></span>
      <!----></span>
    </label></div>
</div>
<div class="fect-radio-group 
            
            ">
  <div class="fect-radio " style="--radioSize: 14px;"><label class=" "><input type="radio" value="small"><span class="fect-radio-name"><span class="fect-radio-point  "></span>
      <!----></span>
    </label></div>
</div>
<div class="fect-radio-group 
            
            ">
  <div class="fect-radio " style="--radioSize: 16px;"><label class=" "><input type="radio" value="medium"><span class="fect-radio-name"><span class="fect-radio-point  "></span>
      <!----></span>
    </label></div>
</div>
<div class="fect-radio-group 
            
            ">
  <div class="fect-radio " style="--radioSize: 18px;"><label class=" "><input type="radio" value="large"><span class="fect-radio-name"><span class="fect-radio-point  "></span>
      <!----></span>
    </label></div>
</div>
`;

exports[`RadioGroup should be support initialValue 1`] = `
<div class="fect-radio-group 
            
            ">
  <div class="fect-radio " style="--radioSize: 16px;"><label class=" "><input type="radio" value="1"><span class="fect-radio-name"><span class="fect-radio-point  active"></span>
      <!----></span>
    </label></div>
  <div class="fect-radio " style="--radioSize: 16px;"><label class=" "><input type="radio" value="2"><span class="fect-radio-name"><span class="fect-radio-point  "></span>
      <!----></span>
    </label></div>
</div>
`;
